From e917c616e2c0fe91e36e430d69488a615d15e393 Mon Sep 17 00:00:00 2001
From: yezengruan <yezengruan@huawei.com>
Date: Wed, 8 Feb 2023 16:03:04 +0800
Subject: [PATCH 500/500] [Huawei] kvm: arm64: fix some pvsched bugs

Offering: HULK
virt inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I6E2VI

--------------------------------------------

1. Clear kvm_vcpu_arch::pvsched::base on vcpu reset
The guest memory will otherwise be corrupted by KVM if we reboot into
an old guest kernel which is **not** aware of the pvsched feature.

2. Fix boot cpu pvsched init abnormal
The pv_sched_init was called too early in the boot in setup_arch, hence
pvsched_vcpu_state was not initializedfo vcpu 0.

Signed-off-by: Zenghui Yu <yuzenghui@huawei.com>
Signed-off-by: yezengruan <yezengruan@huawei.com>
Signed-off-by: Jialin Zhang <zhangjialin11@huawei.com>
---
 arch/arm64/kernel/paravirt.c | 1 +
 arch/arm64/kernel/setup.c    | 2 --
 arch/arm64/kvm/arm.c         | 2 ++
 3 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/arch/arm64/kernel/paravirt.c b/arch/arm64/kernel/paravirt.c
index 45df16e9d8bc..f532534eed53 100644
--- a/arch/arm64/kernel/paravirt.c
+++ b/arch/arm64/kernel/paravirt.c
@@ -351,3 +351,4 @@ int __init pv_sched_init(void)
 
 	return 0;
 }
+early_initcall(pv_sched_init);
diff --git a/arch/arm64/kernel/setup.c b/arch/arm64/kernel/setup.c
index 940e237009e9..0d6cd99f73f5 100644
--- a/arch/arm64/kernel/setup.c
+++ b/arch/arm64/kernel/setup.c
@@ -424,8 +424,6 @@ void __init __no_sanitize_address setup_arch(char **cmdline_p)
 	smp_init_cpus();
 	smp_build_mpidr_hash();
 
-	pv_sched_init();
-
 	/* Init percpu seeds for random tags after cpus are set up. */
 	kasan_init_tags();
 
diff --git a/arch/arm64/kvm/arm.c b/arch/arm64/kvm/arm.c
index 4936fd7b2ddd..5a04f515814e 100644
--- a/arch/arm64/kvm/arm.c
+++ b/arch/arm64/kvm/arm.c
@@ -1123,6 +1123,8 @@ static int kvm_arch_vcpu_ioctl_vcpu_init(struct kvm_vcpu *vcpu,
 	else
 		vcpu->arch.power_off = false;
 
+	kvm_arm_pvsched_vcpu_init(&vcpu->arch);
+
 	return 0;
 }
 
-- 
2.17.1

